### PR TITLE
feat: [PL-42569]: Implement command and outputVar

### PR DESCRIPTION
**Pull request by xingchi-jin**
1. Support command for container steps and shell types
2. Support output var definition
3. Support running lite engine in local and be able to skip cert verify when sending response

Test
1. I executed the bijou based container step with output var abc
image

2. I see the output value of "abc" be successfully send back through task response
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56149)

### CODE REVIEW

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56149)* status: 19 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 748 at r2](https://reviewable.io/reviews/harness/harness-core/56149#-NkpSkMV5GOoDIg4nVYO:-Nkpo-P37IK2bRC4ig7Q:b9jb79r) ([raw file](https://github.com/harness/harness-core/blob/fae8d86fc3ff3a92622d3ff1b183be710b72dc1d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L748)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

in which scenario do we have DELEGATE_GROUP_NAME missing? Is it localdev? We should update the localdev run config instead. 
We almost never use delegateName config (currently no usages), but also GroupName and DelegateName don't necessarily need to match (though they do in 99% of the cases).
</blockquote></details>

this DELEGATE\_GROUP\_NAME is not set in our ng yaml, we set DELEGATE_NAME
the DELEGATE_NAME is set into the delegateName in config file 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/56149#-NkpVCeS9i-r0_FWUxEf:-NkpoaJL4967w45TGbWe:b-rytyyr) ([raw file](https://github.com/harness/harness-core/blob/fae8d86fc3ff3a92622d3ff1b183be710b72dc1d/260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java#L42)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

I think the root cause of your naming constraint issue was using this method directly from `SecretsBuilder.createSecret()`
I think what you should do instead of changing the pattern replacer is make this method private (it's not intended to be used outside this class), and use `getSecretName` instead (in the createSecret). I've made these helper methods all to add a prefix to avoid this exact scenario that you faced. Each helper method would add a different prefix based on the type of resource. This is just a hack of course, but it actually helps when looking for the resources in the K8S (e.g. you want all bijou pods - they all start with `harness-` and and with `-job`
Also, not all the resources have the same naming constraints, so I don't think the pattern replacer should be global.
</blockquote></details>

the error is below
it's the label that is invalid.

our delegate task id uses base64UUID, for base64UUID is has 4 types none alphanumetic chars,  - _ / +    
so replacing all these 4 types chars, and all possible problems can be avoided. 



_Code snippet:_
```Java
Fix bug 2023-11-28 23:11:49,500 [task-exec-8] ERROR io.harness.delegate.service.core.litek8s.SecretsBuilder - K8S ApiException: 422, {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Secret \"secret-mlctgyeyt8obgzogzxasow-del\" is invalid: metadata.labels: Invalid value: \"-ybtheamrhagjlf8zpaj6a-del\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')","reason":"Invalid","details":{"name":"secret-mlctgyeyt8obgzogzxasow-del","kind":"Secret","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: \"-ybtheamrhagjlf8zpaj6a-del\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')","field":"metadata.labels"}]},"code":422}
```
___
*[`420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java` line 169 at r2](https://reviewable.io/reviews/harness/harness-core/56149#-NkpY4E-8RL9mYwLqd9Z:-Nkpp-kj2E_g53kcfRra:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/fae8d86fc3ff3a92622d3ff1b183be710b72dc1d/420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java#L169)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

You shouldn't return full stack trace to the caller. `withDescription` should be just fine (you should log the stacktrace locally though).
</blockquote></details>

Done
___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto` line 25 at r2](https://reviewable.io/reviews/harness/harness-core/56149#-NkpZ81qDpxbkzGMbsEp:-Nkpp1JODaN8evzGuDQ2:b-e1ndnq) ([raw file](https://github.com/harness/harness-core/blob/fae8d86fc3ff3a92622d3ff1b183be710b72dc1d/920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto#L25)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

I think of Command as just another form of an input for a plugin. E.g. plugin input could be a file that you just copy to the container (i.e. legacy tasks converted to containers, using executor lib), a Command or can be other options in the future. Also, these new option you added are specific to the container/run step, but we make them first class objects in our top level resource which is used by the Runner (not LE). Finally I don't think it makes sense to have both ExecutionInput.data and Command at the same time, so I suggest that we move `Command` under the `ExecutionInput` and make it `oneof`. 
Also, output env_var_outputs (remove last `S` btw), is [currently] specific to the task/container step, so I think it should also go inside the Command. 
With these changes, I also think `message Command` could have a better name (but I don't have a good one for you :smile:)
I think this would also help you avoid using UnitStep on the BE, so that we don't create this extra coupling.
</blockquote></details>

Yea, make sense. I will put command into the ExecutionInput.  with one minor difference. i think the command and the inputData is not oneof relationship. the api can provide a flexibility that can support both types of input at the same time. How to pass inputs to tasks is up to the module teams.

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56149)* status: 19 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 748 at r2](https://reviewable.io/reviews/harness/harness-core/56149#-NkpSkMV5GOoDIg4nVYO:-NkrMxRo1XYIr9vn-_mG:bdg8o7c) ([raw file](https://github.com/harness/harness-core/blob/fae8d86fc3ff3a92622d3ff1b183be710b72dc1d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L748)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

this DELEGATE\_GROUP\_NAME is not set in our ng yaml, we set DELEGATE_NAME
the DELEGATE_NAME is set into the delegateName in config file
</blockquote></details>

i removed this change from this PR 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/56149#-NkpVCeS9i-r0_FWUxEf:-NkrN0qB5aVIEmquOsWc:b62yx1p) ([raw file](https://github.com/harness/harness-core/blob/fae8d86fc3ff3a92622d3ff1b183be710b72dc1d/260-delegate/src/main/java/io/harness/delegate/service/core/util/K8SResourceHelper.java#L42)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

the error is below
it's the label that is invalid.

our delegate task id uses base64UUID, for base64UUID is has 4 types none alphanumetic chars,  - _ / +    
so replacing all these 4 types chars, and all possible problems can be avoided.
</blockquote></details>

i removed the change from this PR

**Comment by xingchi-jin** [comment]
trigger uts

File: 260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java [hunk 1]
```diff
@@ -180,49 +179,34 @@ public void init(
...
-    if (isNotEmpty(accountKey)) {
-      executeStepRequest =
-          executeStepRequest.toBuilder().setAccountKey(TokenUtils.getDecodedTokenString(accountKey)).build();
-    }
-    if (isNotEmpty(delegateID)) {
-      executeStepRequest = executeStepRequest.toBuilder().setDelegateId(delegateID).build();
-    }
-
-    final var namespace = config.getNamespace();
+  public void execute(final String infraId, final String logKey, final InputData tasks, Map<String, char[]> decrypted,
```
Comment (gn-harness) [inline 260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java:182]: nit: why is this a plural?

**Review by mdmilic — APPROVED** [review]
> Reviewed 1 of 15 files at r1, 3 of 11 files at r2, 1 of 3 files at r3, 5 of 8 files at r4, 13 of 13 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56149)* status: 28 file(s) to review. 1 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java` line 197 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-NktXsC6D7uaqGcUTdzB:-NktXsC6D7uaqGcUTdzC:b-varl52) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java#L197)):*
> ```Java
>         executeStepBuilder.setTaskParameters(taskData.getBinaryData());
>       }
>       if (Objects.nonNull(k8sExecution.getEntryPoint())) {
> ```

nit: in proto you have hasField (for non primitive types) which is generally considered safer. I think this mattered more in proto2 (because of optional fields), but it's probably good idea to use it. 
___
*[`420-delegate-service/BUILD.bazel` line 31 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-Nktdg59Ayr0zOW73ofX:-Nktdg5A3uutOM8wQfY0:b-z9v4q8) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/420-delegate-service/BUILD.bazel#L31)):*
> ```Starlark
>     ] + [
>         "//955-delegate-beans/src/main/proto/io/harness/delegate/core/beans:955_delegate_core_task_java_proto",
>         "//product/ci/engine/proto:ciengine_java_proto",
> ```

You shouldn't need this dep any more
___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto` line 14 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-NktRDUjCZr6ZcciabPF:-NktRDUjCZr6ZcciabPG:b-y9l11) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto#L14)):*
> ```proto
> 
> message Execution {
>   oneof spec {
> ```

Nice, this is a good overall structure for the proto objects. The only thing I would change is to not use K8S keyword in the naming. Maybe something related to PluginExecution or so, but that's also too generic, I see this as a one specific (not generic) way of executing plugins (unrelated to the type of infra). On the other hand it might be good to tie it to the infra, but then also technically you could have multiple ways to execute on K8S, so not really sure.
Let's merge it as is anyway to unblock other work and we can think separately if we should keep the naming or change it. 
___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto` line 19 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-NktWfvf7OT_WIM4hN0H:-NktWfvf7OT_WIM4hN0I:by9v8gv) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto#L19)):*
> ```proto
>   string infra_ref_id = 2;   // execution infra reference id created by calling SetupExecutionInfrastructureRequest
>   string step_log_key = 3;
>   repeated string env_var_outputs = 4;
> ```

I think these env var output probably belong inside ExecutionSpec

**Comment by xingchi-jin** [comment]
trigger uts

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56149)* status: 24 file(s) to review. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java` line 197 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-NktXsC6D7uaqGcUTdzB:-NkvIi8b6Nxc1sNVAvwm:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/260-delegate/src/main/java/io/harness/delegate/service/core/litek8s/K8SLiteRunner.java#L197)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

nit: in proto you have hasField (for non primitive types) which is generally considered safer. I think this mattered more in proto2 (because of optional fields), but it's probably good idea to use it.
</blockquote></details>

done
___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto` line 14 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-NktRDUjCZr6ZcciabPF:-NkvIjGZ1cdmrmDjaXY3:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto#L14)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Nice, this is a good overall structure for the proto objects. The only thing I would change is to not use K8S keyword in the naming. Maybe something related to PluginExecution or so, but that's also too generic, I see this as a one specific (not generic) way of executing plugins (unrelated to the type of infra). On the other hand it might be good to tie it to the infra, but then also technically you could have multiple ways to execute on K8S, so not really sure.
Let's merge it as is anyway to unblock other work and we can think separately if we should keep the naming or change it.
</blockquote></details>

done
___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto` line 19 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-NktWfvf7OT_WIM4hN0H:-NkvIsbvAYZ90f8meCgm:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/920-delegate-service-beans/src/main/proto/io/harness/delegate/execution.proto#L19)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

I think these env var output probably belong inside ExecutionSpec
</blockquote></details>

done
___
*[`420-delegate-service/BUILD.bazel` line 31 at r5](https://reviewable.io/reviews/harness/harness-core/56149#-Nktdg59Ayr0zOW73ofX:-NkvIuJE6wwFKz20DPVM:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5b8b732fa3726eeca3fb0fb09f07e5856c422504/420-delegate-service/BUILD.bazel#L31)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

You shouldn't need this dep any more
</blockquote></details>

done

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger ut9

**Comment by xingchi-jin** [comment]
trigger ut9

**Comment by xingchi-jin** [comment]
Test 4 and test 6 each has a test failing, not related to this PR. I have posted in #help-engineering and tagged the owner to fox.

**Pull request merged**